test/types/array.js

```diff
@@ -578,13 +578,73 @@ describe('array', () => {
 
         it('can fill the default values into the value array', () => {
 
-            const schema = Joi.array().ordered( Joi.string().required(), Joi.number().default(0), Joi.number().default(6)).required();
+            const schema = Joi.array().ordered(Joi.string().required(), Joi.number().default(0), Joi.number().default(6)).required();
 
             Helper.validate(schema, [
                 [['a'], true, ['a', 0, 6]]
             ]);
         });
 
+        it('ignore trailing undefined', () => {
+
+            const schema = Joi.array().ordered(Joi.string().required(), Joi.string(), Joi.number().default(5), Joi.string(), Joi.string());
+
+            Helper.validate(schema, [
+                [['a'], true, ['a', undefined, 5]]
+            ]);
+        });
+
+        it('generate sparse array', () => {
+
+            const schema = Joi.array().ordered(Joi.string(), Joi.number().default(5), Joi.string(), Joi.string().default('f'));
+
+            Helper.validate(schema, [
+                [[], true, [undefined, 5, undefined, 'f']]
+            ]);
+        });
+
+        it('fill the defaults correctly when dynamic schema is used', () => {
+
+            const schema = Joi.object({
+                info: Joi.object(),
+                values: Joi.array().ordered(
+                    Joi.string().required(),
+                    Joi.string().default(Joi.ref('info.firstname')),
+                    Joi.string(),
+                    Joi.string().default(Joi.ref('info.lastname')),
+                    Joi.string()
+                ).required()
+            });
+
+            Helper.validate(schema, [
+                [{ info: { firstname: 'f', lastname: 'e' }, values: ['h'] }, true, { info: { firstname: 'f', lastname: 'e' }, values: ['h', 'f', undefined, 'e'] }],
+                [{ info: { firstname: 'f', lastname: 'e' }, values: ['h', 'test'] }, true, { info: { firstname: 'f', lastname: 'e' }, values: ['h', 'test', undefined, 'e'] }]
+            ]);
+        });
+
+        it('fill the defaults correctly when dynamic schema is used', () => {
+
+            const schema = Joi.object({
+                info: Joi.object().required(),
+                values: Joi.array().ordered(
+                    Joi.string().required(),
+                    Joi.array().ordered(
+                        Joi.string(),
+                        Joi.string().default('normal value'),
+                        Joi.string(),
+                        Joi.number().default(Joi.x('{number("202099")}'))
+                    ),
+                    Joi.string(),
+                    Joi.string().default(Joi.ref('info.firstname')),
+                    Joi.string()
+                ).required()
+            });
+
+            Helper.validate(schema, [
+                [{ info: { firstname: 'f' }, values: ['h', []] }, true, { info: { firstname: 'f' }, values: ['h', [undefined, 'normal value', undefined, 202099], undefined, 'f'] }]
+            ]);
+        });
+
```

Comment:
A test case for single value converted to array would be helpful just to make sure the behaviour is consistent. Also case that checks if errors are returned before filling is also needed

Reply:
What do you mean by a single value converted to array?

Reply:
@iifawzi There's `array.single()` which basically wraps single values in arrays. For example, `2` becomes `[2]`. In this case, the path included in error reports should exclude its index (because the original value wasn't in an array)

Reply:
Ignore that. Because we are sure that no errors are going to be returned, there is no need. You don't even need the check `wasArray` because it's only to exclude paths of single values as explained. Test `Joi.link()` instead.

Reply:
So now the tests that needed to be added is just `Joi.link()`. 
and for the above suggestion it will only be as following: 
```JS
const ancestors = [value, ...state.ancestors];
        const ordered = ordereds[i];
        const res = ordered.$_validate(undefined, state.localize(state.path, ancestors, ordered), prefs);
```
are we ok with this?

Reply:
@iifawzi I'm not sure if `state.path` is only used for error generation. I have a feeling it might be used by `Joi.link()` to resolve schemas. If it does, you need to localize the path as well.

Reply:
Stuck. Can you help me writing a schema that can test `Joi.link()` with ordered array? as unfortunately i'm not able to think about any.

Reply:
@iifawzi You don't need to think of one now. I'm not very familiar with `Joi.link()` either. I will have to dive into it sometime tomorrow.

Reply:
For an hour, i was thinking about a schema that could. be used, i couldn't figure one.
all other mentioned points are resolved, i will commit them for now.